### ISSUE TITLE
Build failure: elfutils (i686-linux)

### ISSUE THREAD

**Issue by someplaceguy** [NixOS/nixpkgs#329218]
### Steps To Reproduce

1. `nix-build '<nixpkgs>' -A pkgsi686Linux.elfutils`

Tested with `nixos-24.05` branch (commit 63d37ccd2d178d54e7fb691d7ec76000740ea24a).

### Build log

Build fails in `checkPhase`:

```
FAIL: run-backtrace-native.sh
=============================

case 0: expected symname 'raise' got '_dl_sysinfo_int80'
./test-subr.sh: line 84: 12197 Aborted                 (core dumped) LD_LIBRARY_PATH="${built_library_path}${LD_LIBRARY_PATH:+:}$LD_LIBRARY_PATH" $VALGRIND_CMD "$@"
backtrace-child: no main
FAIL run-backtrace-native.sh (exit status: 1)
```

### Notify maintainers

cc @edolstra @r-burns 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.40, NixOS, 24.05 (Uakari), 24.05.git.b54eb1bb552c`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.4`
 - nixpkgs: `/nixdev/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by trofi** [comment]
Do you know why you don't get it from a substitute server? It seems to build fine on hydra.

Backtrace breakage looks a bit worrying. Any unusual setup you might have done that could break it?

For me the test passes as `PASS: run-backtrace-native.sh`

**Comment by someplaceguy** [comment]
> Do you know why you don't get it from a substitute server? 

I don't get it from a substitute server because I've disabled substitute servers. I prefer building packages locally.

> It seems to build fine on hydra.

@trofi Where can I see the Hydra build log for `pkgsi686Linux.elfutils` on the 24.05 branch?

My search only revealed builds for x86_64-linux and aarch64-linux: https://hydra.nixos.org/eval/1807987?filter=elfutils&compare=1807954&full=

Another search also revealed this build, but this is for the master branch, not 24.05: https://hydra.nixos.org/job/nixpkgs/trunk/elfutils.i686-linux

> Any unusual setup you might have done that could break it?

No, nothing I can think of...

**Comment by trofi** [comment]
If we pick latest eval of, say, `wine`: https://hydra.nixos.org/job/nixos/release-24.05/nixpkgs.wine.x86_64-linux and look at Build Dependencies tab this would be an example build: https://hydra.nixos.org/build/266242457/nixlog/37

You can try to diff your log against it to see where the discrepancy crops up. Might be related to the abilty to collect crash dumps.

**Comment by someplaceguy** [comment]
> You can try to diff your log against it to see where the discrepancy crops up.

The log diff did not reveal anything substantial, but fortunately I was able to figure out what the problem was.

> Might be related to the abilty to collect crash dumps.

I've looked into the test and checked that the crash dumps / stack traces were actually working fine.

Here's the long explanation of the root cause of this issue (and a TL;DR summary at the end):

# Long explanation

The test that was failing was performing a `raise()` call, then doing a stack trace, then checking to see if the stack trace contained what was expected.

The test failed because it found an unexpected call to a function called `_dl_sysinfo_int80` within the stack trace, in-between the calll to `raise()` and the kernel entry-point. What is this `_dl_sysinfo_int80` function? It's just a function that performs the `int 0x80` instruction, which is the old and slow way of performing a system call in i686, which works on all x86-32 CPUs (and x86-64 CPUs in 32-bit mode).

Why was this C program in the elfutils test suite using the old and slow way of performing system calls instead of the new, faster way that modern CPUs support? Well, glibc by default uses the slow way of performing system calls for compatibility reasons, unless the kernel provides special code to perform these system calls. This mechanism of the kernel providing special code to userspace is called [vDSO](https://en.wikipedia.org/wiki/VDSO).

Why was my kernel not providing this special code to userspace? Well, that's because I was using the `vdso32=0` kernel command-line parameter. Why was I using this kernel parameter? Because [it was recommended in this page](https://kspp.github.io/Recommended_Settings#x86_64-1), for security reasons.

It turns out this kernel command-line parameter is a bit overly-aggressive. The page above recommends it just to make sure that `COMPAT_VDSO` (which is a now-useless kernel option with questionable security implications) is really disabled, but since my kernel config disables this option and nixpkgs already makes sure that the resulting kernel config options match what were really configured, it's not really necessary to completely disable vDSO for 32-bit applications just to make sure that `COMPAT_VDSO` is not accidentally enabled.

# TL;DR summary

The [kernel security hardening wiki page](https://kspp.github.io/Recommended_Settings#x86_64-1) recommended the kernel command-line parameter `vdso32=0`, but 1) it turns out that this parameter is unnecessary if `COMPAT_VDSO` is disabled in the kernel config, and 2) this parameter makes system calls for 32-bit applications slower, and 3) this parameter breaks the elfutils test because elfutils was not expecting a test program to be using the old, slow way of performing system calls.

Anyway, I will now close this issue since elfutils now builds fine for me after removing the `vdso32=0` kernel parameter.

**Comment by trofi** [comment]
Yeah, disabling `vdso` would do it. Might also be worthwhile reporting test failures to `elfutils` upstream. I think it's reasonable to expect tests to Just Work both with and without `vdso` presence (even if it causes some performance degradation on real use cases).
